internal/consumer/partitionConsumer.go

```diff
@@ -183,11 +166,20 @@ func (p *partitionConsumer) markOffset() {
 
 // deliver delivers a message through the consumer channel
 func (p *partitionConsumer) deliver(scm *sarama.ConsumerMessage) {
+	metadata := newDLQMetadata()
+	if p.decodeDLQMetadata {
+		if err := proto.Unmarshal(scm.Key, metadata); err != nil {
+			p.logger.Fatal("unable to marshal dlq metadata from message key", zap.Error(err))
+			return
+		}
+	}
+
 	ackID, err := p.trackOffset(scm.Offset)
 	if err != nil {
 		return
 	}
-	msg := newMessage(scm, ackID, p.ackMgr, p.dlq)
+
+	msg := newMessage(scm, ackID, p.ackMgr, p.dlq, *metadata)
```

Comment:
instead of passing an additional arg and adding a bunch of ifs in the Message, how about simply overwriting the values in the original message when metadata is non-nil ? Once the ackID is built, there is no need for original metadata ?

Reply:
I have found (in our internal implementation) that it is useful for debugging (both for the lib owner and the user of the lib) to retain all of the information.

Reply:
but in your current implementation, doesn't look like there is any way to retrieve the original information, is there ?

Reply:
There isn't. But it would be trivial to include `zapcore.ObjectMarshaler` to `kafka.Message` interface so that you can do something like `log.Info("kafka message", zap.Object("kafkaMsg", msg)`.